components/salsa-2022/src/input.rs

```diff
@@ -49,9 +49,12 @@ where
         Id::from_id(crate::Id::from_u32(next_id))
     }
 
-    pub fn new_singleton_input(&mut self, _runtime: &Runtime) -> Id {
-        // There's only one singleton so record that we've created it
-        // and return the only id.
+    pub fn new_singleton_input(&self, _runtime: &Runtime) -> Id {
+        // when one exists already, panic
+        if self.counter.load(Ordering::Relaxed) >= 1 {
+            panic!("Singleton struct may not be duplicated");
```

Comment:
```suggestion
            panic!("singleton struct may not be duplicated");
```

For some reason, starting with a lowercase letter seems to fit Rust style more to my eyes. Not sure if there's any official guidance on this point.